README.adoc

```diff
@@ -0,0 +1,402 @@
+    this.height = height
+    this.width = width
+  }
+}
+```
+
+**Pass**
+
+```js
+const polygon = (height, width) => ({
+  height: height,
+  width: width,
+})
+```
+
+### `fp/no-delete`
+
+> delete is an operator to remove fields from an object or elements from an array. This purposely mutates data, which is not wanted when doing functional programming.
+
+Further reading: https://github.com/google/google-api-nodejs-client/issues/375[Avoid using delete operator]
+
+**Fail**
+```js
+delete foo
+delete foo.bar
+delete foo[bar]
+```
+**Pass**
+```js
+import F from "ramda"
+
+const fooWithoutBar = F.omit(["bar"], foo)
+const fooWithoutField = F.omit([bar], foo)
+```
+
+### `fp/no-events`
+
+> The use of EventEmitter with the events module provided by Node.js promotes implicit side-effects by emitting and listening to events. Instead of events, you should prefer activating the wanted effects by calling the functions you wish to use explicitly.
+
+Probably what you should do is use a https://gist.github.com/staltz/868e7e9bc2a7b8c1f754[functional reactive programming] library: https://github.com/cujojs/most[`most`], https://github.com/Reactive-Extensions/RxJS[`rxjs`].
+
+**Fail**
+
+```js
+import EventEmitter from "events"
+```
+
+### `fp/no-get-set`
+
+**Fail**
+```js
+const person = {
+  name: 'Some Name',
+  get age () {
+    return this._age
+  },
+  set age (n) {
+    if (n < 0) {
+      this._age = 0
+    } else if (n > 100) {
+      this._age = 100
+    } else {
+      this._age = n
+    }
+  }: 20
+};
+
+person.__defineGetter__("name", function () {
+  return this.name || "John Doe";
+})
+
+person.__defineSetter__("name", function (name) {
+  this.name = name.trim();
+})
+```
+**Pass**
+```js
+import F from "ramda"
+
+const person = {
+  name: "Some Name",
+  age: 20,
+}
+
+const clamp = (n, min, max) =>
+  n <= min ? min :
+  n >= max ? max :
+             n
+
+const setAge = (age, person) =>
+  F.merge(person, { age: clamp(age, 0, 100) })
+```
+
+### `fp/no-let`
+
+> If you want to program as if your variables are immutable, part of the answer is to not allow your variables to be reassigned. By not allowing the use of let and var, variables that you declared may not be reassigned.
+
+**Fail**
+```js
+let a = 1
+let b = 2,
+    c = 3
+let d
+```
+**Pass**
+```js
+const a = 1
+const b = 2,
+      c = 3
+```
+
+### `fp/no-loops`
+> Loops, such as for or while loops, work well when using a procedural paradigm. In functional programming, recursion or implementation agnostic operations like map, filter and reduce are preferred.
+
+**Fail**
+```js
+const result = []
+const elements = [1, 2, 3]
+
+for (let i = 0; i < elements.length; i++) {
+  if (elements[i] > 2) {
+    result.push(elements[i])
+  }
+}
+
+for (element in elements) {
+  result.push(element * 10)
+}
+
+while (n < 100) {
+  result.push(n)
+  n *= 2
+}
+```
+**Pass**
+```js
+const xs = [1, 2, 3]
+
+xs.filter(x => x > 2)
+
+xs.map(x => x * 10)
+
+const doubleBlast (n) =>
+  n >= 100
```

Comment:
Below is a real world example showing just one case of how statements lead to sweeping damage on readability. Even if record creation had not been factored out, it would still be a significant improvement.

##### Before
```js
  checkES() {
    return this.client.cluster.health({
      level: "indices",
    })
      .then((health) => {
        const allowedStatuses = ["green", "yellow"]
        let resp = null
        // check cluster health
        if (allowedStatuses.indexOf(health.status) === -1) {
          resp = {
            code: 500,
            message: "Cluster did not pass healthcheck",
            health,
          }
        } else {
          resp = {
            code: 200,
            message: "Cluster is doing grrreat!",
            health,
          }
        }
        return resp
      })
  }
}
```

#### After
```js
const checkES = client => (
  client.cluster
  .health({ level: "indices" })
  .then(health => (
    F.contains(health.status, ["green", "yellow"])
      ? Records.createHealthy(health)
      : Records.createUnhealthy(health)
  ))
)
```